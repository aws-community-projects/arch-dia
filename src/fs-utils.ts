import * as fs from 'fs';

function writeFile (file: string, val: string): void {
  fs.writeFileSync(file, val);
}

function readFile (file: string): string {
  return fs.readFileSync(file, 'utf8');
}

export { writeFile, readFile };
