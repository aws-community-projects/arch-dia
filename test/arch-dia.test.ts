import { App, Aspects, Stack } from 'aws-cdk-lib';
import { Queue } from 'aws-cdk-lib/aws-sqs';

const writeFile = jest.fn();
const readFile = jest.fn();
jest.mock('../src/fs-utils', () => ({
  readFile,
  writeFile,
}));

// eslint-disable-next-line import/first
import { ArchitectureDiagramAspect } from '../src';

describe('ArchitectureDiagramAspect', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  test('Basic Stack', () => {
    readFile.mockReturnValueOnce(''); // mock empty file
    const app = new App();
    const stack = new Stack(app, 'MyTestStack');
    const archDiagramAspect = new ArchitectureDiagramAspect();
    archDiagramAspect.visit(stack);

    const dia = archDiagramAspect.generateDiagram();

    expect(dia).toMatchSnapshot();
    expect(readFile).toHaveBeenCalledTimes(1);
    expect(writeFile).toHaveBeenCalledTimes(1);
  });

  test('Add Queue to Stack', () => {
    readFile.mockReturnValueOnce(''); // mock empty file
    const initialApp = new App();
    const initialStack = new Stack(initialApp, 'MyTestStack');
    const initialAspect = new ArchitectureDiagramAspect();
    initialAspect.visit(initialStack);
    const initialDiagram = initialAspect.generateDiagram();
    readFile.mockReturnValueOnce(initialDiagram);
    const updatedApp = new App();
    const updatedStack = new Stack(updatedApp, 'MyTestStack');
    const updatedAspect = new ArchitectureDiagramAspect();
    new Queue(updatedStack, 'MyQueue');
    updatedAspect.visit(updatedStack);
    const updatedDiagram = updatedAspect.generateDiagram();
    expect(updatedDiagram).toMatchSnapshot();

    expect(writeFile).toHaveBeenCalledTimes(2);
  });

  test('Normalize after addition', () => {
    readFile.mockReturnValueOnce(''); // mock empty file
    const initialApp = new App();
    const initialStack = new Stack(initialApp, 'MyTestStack');
    const initialAspect = new ArchitectureDiagramAspect();
    initialAspect.visit(initialStack);
    const initialDiagram = initialAspect.generateDiagram();
    readFile.mockReturnValueOnce(initialDiagram);
    const updatedApp = new App();
    const updatedStack = new Stack(updatedApp, 'MyTestStack');
    const updatedAspect = new ArchitectureDiagramAspect();
    new Queue(updatedStack, 'MyQueue');
    updatedAspect.visit(updatedStack);
    const updatedDiagram = updatedAspect.generateDiagram();
    readFile.mockReturnValueOnce(updatedDiagram);
    expect(updatedAspect.generateDiagram()).toMatchSnapshot();

    expect(writeFile).toHaveBeenCalledTimes(3);
  });

  test('Remove Queue from Stack', () => {
    readFile.mockReturnValueOnce(''); // mock empty file
    const initialApp = new App();
    const initialStack = new Stack(initialApp, 'MyTestStack');
    const initialAspect = new ArchitectureDiagramAspect();
    new Queue(initialStack, 'MyQueue');
    initialAspect.visit(initialStack);
    const initialDiagram = initialAspect.generateDiagram();
    readFile.mockReturnValueOnce(initialDiagram);
    const updatedApp = new App();
    const updatedStack = new Stack(updatedApp, 'MyTestStack');
    const updatedAspect = new ArchitectureDiagramAspect();
    updatedAspect.visit(updatedStack);
    const updatedDiagram = updatedAspect.generateDiagram();
    expect(updatedDiagram).toMatchSnapshot();

    expect(writeFile).toHaveBeenCalledTimes(2);
  });

  test('Normalize after removal from Stack', () => {
    readFile.mockReturnValueOnce(''); // mock empty file
    const initialApp = new App();
    const initialStack = new Stack(initialApp, 'MyTestStack');
    const initialAspect = new ArchitectureDiagramAspect();
    new Queue(initialStack, 'MyQueue');
    initialAspect.visit(initialStack);
    const initialDiagram = initialAspect.generateDiagram();
    readFile.mockReturnValueOnce(initialDiagram);
    const updatedApp = new App();
    const updatedStack = new Stack(updatedApp, 'MyTestStack');
    const updatedAspect = new ArchitectureDiagramAspect();
    updatedAspect.visit(updatedStack);
    const updatedDiagram = updatedAspect.generateDiagram();
    readFile.mockReturnValueOnce(updatedDiagram);
    expect(updatedAspect.generateDiagram()).toMatchSnapshot();

    expect(writeFile).toHaveBeenCalledTimes(3);
  });
});
