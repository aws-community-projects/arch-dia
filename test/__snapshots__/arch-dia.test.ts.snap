// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArchitectureDiagramAspect Add Queue to Stack 1`] = `
"# MyTestStack

\`\`\`mermaid
graph LR;
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack]
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack] ==> c8a33fea56cf1bd6e8b5c652f393a6d315e7a78836[MyQueue]:::added
c8a33fea56cf1bd6e8b5c652f393a6d315e7a78836[MyQueue]:::added ==> c80cb3f181e7dbb6726caf602717b2585dabff5aff[Resource - AWS::SQS::Queue]:::added
classDef default fill:#fff,stroke:#000,color:black;
classDef added fill:#cfc,stroke:#cfc,stroke-width:2px,color:black;
classDef removed fill:#fcc,stroke:#fcc,stroke-width:2px,color:black;
\`\`\`
"
`;

exports[`ArchitectureDiagramAspect Basic Stack 1`] = `
"# MyTestStack

\`\`\`mermaid
graph LR;
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack]
classDef default fill:#fff,stroke:#000,color:black;
classDef added fill:#cfc,stroke:#cfc,stroke-width:2px,color:black;
classDef removed fill:#fcc,stroke:#fcc,stroke-width:2px,color:black;
\`\`\`
"
`;

exports[`ArchitectureDiagramAspect Normalize after addition 1`] = `
"# MyTestStack

\`\`\`mermaid
graph LR;
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack]
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack] --> c8a33fea56cf1bd6e8b5c652f393a6d315e7a78836[MyQueue]
c8a33fea56cf1bd6e8b5c652f393a6d315e7a78836[MyQueue] --> c80cb3f181e7dbb6726caf602717b2585dabff5aff[Resource - AWS::SQS::Queue]
classDef default fill:#fff,stroke:#000,color:black;
classDef added fill:#cfc,stroke:#cfc,stroke-width:2px,color:black;
classDef removed fill:#fcc,stroke:#fcc,stroke-width:2px,color:black;
\`\`\`
"
`;

exports[`ArchitectureDiagramAspect Normalize after removal from Stack 1`] = `
"# MyTestStack

\`\`\`mermaid
graph LR;
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack]
classDef default fill:#fff,stroke:#000,color:black;
classDef added fill:#cfc,stroke:#cfc,stroke-width:2px,color:black;
classDef removed fill:#fcc,stroke:#fcc,stroke-width:2px,color:black;
\`\`\`
"
`;

exports[`ArchitectureDiagramAspect Remove Queue from Stack 1`] = `
"# MyTestStack

\`\`\`mermaid
graph LR;
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack]
c814b2c5dc22d11a84e78a63be7bb7815715b70534[MyTestStack] -.-> c8a33fea56cf1bd6e8b5c652f393a6d315e7a78836[MyQueue]:::removed
c8a33fea56cf1bd6e8b5c652f393a6d315e7a78836[MyQueue]:::removed -.-> c80cb3f181e7dbb6726caf602717b2585dabff5aff[Resource - AWS::SQS::Queue]:::removed
classDef default fill:#fff,stroke:#000,color:black;
classDef added fill:#cfc,stroke:#cfc,stroke-width:2px,color:black;
classDef removed fill:#fcc,stroke:#fcc,stroke-width:2px,color:black;
\`\`\`
"
`;
